ambry-tools/src/main/java/com.github.ambry/tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,956 @@
+            logger.error(failureCount.get() + "  failures during this batch, hence exiting the Put Thread");
+            break;
+          }
+          Thread.sleep(threadLocalRandom.nextLong(sleepTimeBetweenBatchPutsInMs));
+        } catch (InterruptedException ex) {
+          logger.error("InterruptedException in putThread " + ex.getStackTrace());
+        }
+      }
+      putComplete.set(true);
+      logger.info(
+          "Exiting Producer. Current Put Count " + currentPutCount.get() + ", total Put Count " + totalPutCount);
+    }
+  }
+
+  /**
+   * Record to hold information about a blob along with its content and its reference count
+   */
+  class BlobAccessInfo {
+    String blobId;
+    AtomicInteger referenceCount;
+    AtomicInteger burstGetReferenceCount;
+    byte[] blobContent;
+
+    public BlobAccessInfo(String blobId, byte[] blobContent) {
+      this.blobId = blobId;
+      this.referenceCount = new AtomicInteger(0);
+      this.burstGetReferenceCount = new AtomicInteger(0);
+      this.blobContent = blobContent;
+    }
+  }
+
+  /**
+   * Get threads to assist in consuming blobs from ambry and verifying its content
+   */
+  class GetThread implements Runnable {
+    private final PutGetHelper getHelper;
+    private final CopyOnWriteArrayList<BlobAccessInfo> generatedBlobAccessInfos;
+    private final int maxGetCountPerBlob;
+    private final AtomicInteger getCompletedCount;
+    private final AtomicInteger totalPutCount;
+    private final int maxBurstCountPerblob;
+    private final int parallelGetCount;
+    private final boolean deleteOnExit;
+    private final int maxFailuresPerGetBatch;
+    private final long sleepTimeBetweenBatchGetsInMs;
+    private final ThreadLocalRandom threadLocalRandom;
+    private final BlockingQueue<String> deleteBlobIds;
+
+    public GetThread(PutGetHelper getHelper, CopyOnWriteArrayList<BlobAccessInfo> generatedBlobAccessInfos,
+        int maxGetCountPerBlob, final AtomicInteger totalPutCount, AtomicInteger getCompletedCount,
+        int maxBurstCountPerBlob, int parallelGetCount, int maxFailuresPerGetBatch, long sleepTimeBetweenBatchGetsInMs,
+        boolean deleteOnExit, ThreadLocalRandom threadLocalRandom, BlockingQueue<String> deleteBlobIds) {
+      this.getHelper = getHelper;
+      this.generatedBlobAccessInfos = generatedBlobAccessInfos;
+      this.maxGetCountPerBlob = maxGetCountPerBlob;
+      this.totalPutCount = totalPutCount;
+      this.parallelGetCount = parallelGetCount;
+      this.getCompletedCount = getCompletedCount;
+      this.maxBurstCountPerblob = maxBurstCountPerBlob;
+      this.deleteOnExit = deleteOnExit;
+      this.maxFailuresPerGetBatch = maxFailuresPerGetBatch;
+      this.sleepTimeBetweenBatchGetsInMs = sleepTimeBetweenBatchGetsInMs;
+      this.threadLocalRandom = threadLocalRandom;
+      this.deleteBlobIds = deleteBlobIds;
+    }
+
+    @Override
+    public void run() {
```

Comment:
could you add javadoc here or at the beginning of the class describing the intention of this thread? It is not immediately clear why you are executing this while loop only as long as puts are happening.